README.markdown
rit-client
==========

rit-client is a Rails plugin that provides access to the [Rit. content scheduling system](http://github.com/briandoll/Rit/).

Please see the [Rit documentation and wiki](http://github.com/briandoll/Rit/) for more information on Rit.

Install
=======

    ./script/plugin install git://github.com/briandoll/rit_client.git

Configure
=========

    # In config/initializers/rit.rb
    RIT_HOST = 'rit.example.com'
    RIT_PORT = 8080

Basic Usage
===========

Once the plugin is installed you'll be able to access Rit. content from your controllers and views.

A safe method that returns nil if no content is found:

    rit_plate(layout_name, instance_name, plate_name)

An 'unsafe' method that may throw either a Rit::NotFoundError or a Rit::TimeoutError

    rit_plate!(layout_name, instance_name, plate_name)

Usage with Rit
==============

See the [Rit wiki](http://wiki.github.com/briandoll/Rit/) for detailed usage with Rit.


Copyright (c) 2009, 2010 Brian Doll, Kasima Tharnpipitchai, Sheet Music Plus LLC, released under the MIT license


HISTORY.markdown
rit_client 0.2 / 2010-03
========================
 Public Release

rit_client 0.1 / 2009-11
========================
 Designed by Brian Doll and Kasima Tharnpipitchai
 Developed by Kasima Tharnpipitchai
 Serving production content for Sheet Music Plus @ http://sheetmusicplus.com/

test/test_helper.rb
require 'rubygems'
require 'active_support'
require 'active_support/test_case'
require 'shoulda'
require 'rit_client'

test/rit_client_test.rb
require File.dirname(__FILE__) + '/test_helper'

class RitClientTest < Test::Unit::TestCase
  
  context "The Rit::Plate class" do
    should "have a default timeout of 5 seconds" do
      assert_equal(5, Rit::Plate.timeout)
    end

    should "have a different timeout when set" do
      Rit::Plate.timeout = 3
      assert_equal(3, Rit::Plate.timeout)
    end

    should "return a proper rit publish path with an instance name" do
      path = Rit::Plate.published_plate_path('a', 'b', 'c')
      assert_equal('/published/a/b/c', path)
    end

    should "return a proper rit publish path without an instance name" do
      path = Rit::Plate.published_plate_path('a', nil, 'c')
      assert_equal('/published/a/c', path)
    end
  end

  context "Rit::Plate#handle_response" do
    setup do
      @response = mock
    end

    context "when sent a 200 response" do
      setup { @response.stubs(:code).returns('200') }
      should "return response" do
        assert_equal(@response, Rit::Plate.handle_response(@response))
      end
    end

    context "when sent a 404 response" do
      setup { @response.stubs(:code).returns('404') }
      should "raise NotFoundError" do
        assert_raise(Rit::NotFoundError) { Rit::Plate.handle_response(@response) }
      end
    end

    context "when sent a 500 response" do
      setup { @response.stubs(:code).returns('500') }
      should "raise NotFoundError" do
        assert_raise(Rit::ConnectionError) { Rit::Plate.handle_response(@response) }
      end
    end
  end
end

lib/rit_client.rb
require 'net/http'
require 'uri'

module Rit

  class Plate
    DEFAULT_TIMEOUT = 5

    class << self

      def timeout
        if defined?(@timeout)
          @timeout
        else
          @timeout = defined?(RIT_TIMEOUT) ? RIT_TIMEOUT : DEFAULT_TIMEOUT
        end
      end

      def timeout=(val)
        @timeout = val
      end

      def get(layout_name, instance_name, plate_name, publish_time = nil)
        path = published_plate_path(layout_name, instance_name, plate_name, publish_time)
        Rails.logger.info "Publish from rit.: #{path.to_s}"
        res = http.send(:get, path)
        # ms = Benchmark.ms { res = http.send(:get, url.path) }
        # logger.info "--> %d %s (%d %.0fms)" % [result.code, result.message, result.body ? result.body.length : 0, ms]
        handle_response(res).body
      rescue Timeout::Error => e
        raise TimeoutError.new(e.message)
      end

      def published_plate_path(layout_name, instance_name, plate_name, publish_time = nil)
        if publish_time.nil? or RAILS_ENV == 'production'
          '/published/' + [layout_name, instance_name, plate_name].compact.join('/')
        else
          '/published_on/' + [layout_name, instance_name, plate_name, publish_time.to_i].compact.join('/')
        end
      end

      def http
        configure_http(new_http)
      end

      def new_http
        Net::HTTP.new(RIT_HOST, RIT_PORT)
      end

      def configure_http(http)
        # Net::HTTP timeouts default to 60 seconds.
        http.open_timeout = timeout
        http.read_timeout = timeout
        http
      end

      # Handles response and error codes from the remote service.
      def handle_response(response)
        case response.code.to_i
          when 200...400
            response
          when 404
            raise NotFoundError.new(response)
          else
            raise ConnectionError.new(response)
        end
      end
    end
  end


  class ConnectionError < StandardError
    def initialize(response=nil)
      @response = response
    end

    def to_s
      message = "Failed."
      unless @response.nil?
        message << "  Response code = #{@response.code}." if @response.respond_to?(:code)
        message << "  Response message = #{@response.message}." if @response.respond_to?(:message)
      end
      message
    end
  end

  class NotFoundError < ConnectionError; end
  class TimeoutError < ConnectionError; end


  module ControllerMethods

    # A safe method that attempts to render content from rit
    #
    # Arguments:
    # * layout_name: +String+
    #     The name of the rit layout
    # * instance_name: +String+
    #     The name of the rit instance (optional)
    # * plate_name: +String+
    #     The name of the rit plate
    #
    # Returns:
    #   Returns content or empty (string)
    def rit_plate(layout_name, instance_name, plate_name)
      Rit::Plate.get(layout_name, instance_name, plate_name, session[:preview_time])
    rescue Rit::ConnectionError, Rit::NotFoundError
      ''
    end

    # A method that attempts to render content from rit and if none found will raise an error
    #
    # Arguments:
    # * layout_name: +String+
    #     The name of the rit layout
    # * instance_name: +String+
    #     The name of the rit instance (optional)
    # * plate_name: +String+
    #     The name of the rit plate
    #
    # Returns:
    #   Returns content (string) or raises error if not found
    def rit_plate!(layout_name, instance_name, plate_name)
      Rit::Plate.get(layout_name, instance_name, plate_name, session[:preview_time])
    end
  end
end